apps/frontend/src/lib/psi/index.ts

```diff
@@ -22,20 +23,81 @@ export const generatePSIKeyPair = async () => {
-  bitVector[0] = data.desiredConnections.getHealthy ? 1 : 0;
-  bitVector[1] = data.desiredConnections.cowork ? 1 : 0;
-  bitVector[2] = data.desiredConnections.enjoyMeals ? 1 : 0;
-  bitVector[3] = data.desiredConnections.learnFrontierTopics ? 1 : 0;
-  bitVector[4] = data.desiredConnections.findCollaborators ? 1 : 0;
-  bitVector[5] = data.desiredConnections.goExploring ? 1 : 0;
-  bitVector[6] = data.desiredConnections.party ? 1 : 0;
-  bitVector[7] = data.desiredConnections.doMentalWorkouts ? 1 : 0;
-
-  return bitVector;
+const PSI_BIT_VECTOR_SIZE = 10000;
+const COMMON_CONNECTIONS_BIT_VECTOR_SIZE = 9000;
+
+export const generateBitVectorFromUserData = (
+  userData: UserData,
+  connections: Record<string, Connection>
+): { bitVector: Uint32Array; indexMapping: Record<number, string[]> } => {
+  const bitVector = new Uint32Array(PSI_BIT_VECTOR_SIZE).fill(0);
+  const bitIndexToUsername: Record<number, string[]> = {};
+
+  // 0-8999 reserved for common connections
+  for (const connection of Object.values(connections)) {
+    const { username, signaturePublicKey } = connection.user;
+    const hash = sha256(signaturePublicKey);
+    const bigIntHash = BigInt("0x" + hash);
+    const index = Number(
+      bigIntHash % BigInt(COMMON_CONNECTIONS_BIT_VECTOR_SIZE)
+    );
+    bitVector[index] = 1;
+
+    if (!bitIndexToUsername[index]) {
+      bitIndexToUsername[index] = [];
+    }
+    bitIndexToUsername[index].push(username);
+  }
+
+  // 9000-9999 reserved for edge city lanna data
+  const lannaDesiredConnections = userData.lanna?.desiredConnections;
+  if (lannaDesiredConnections) {
+    bitVector[9000] = lannaDesiredConnections.getHealthy ? 1 : 0;
+    bitVector[9001] = lannaDesiredConnections.cowork ? 1 : 0;
+    bitVector[9002] = lannaDesiredConnections.enjoyMeals ? 1 : 0;
+    bitVector[9003] = lannaDesiredConnections.learnFrontierTopics ? 1 : 0;
+    bitVector[9004] = lannaDesiredConnections.findCollaborators ? 1 : 0;
+    bitVector[9005] = lannaDesiredConnections.goExploring ? 1 : 0;
+    bitVector[9006] = lannaDesiredConnections.party ? 1 : 0;
+    bitVector[9007] = lannaDesiredConnections.doMentalWorkouts ? 1 : 0;
+  }
+
+  return { bitVector, indexMapping: bitIndexToUsername };
+};
+
+export const getOverlapFromPSIResult = (
+  overlapIndices: number[],
+  indexMapping: Record<number, string[]>
+): PSIData => {
+  const sharedConnections: string[] = [];
+  const sharedLannaDesiredConnections: string[] = [];
+
+  for (const index of overlapIndices) {
+    if (index < COMMON_CONNECTIONS_BIT_VECTOR_SIZE) {
+      if (indexMapping[index]) {
+        sharedConnections.push(...indexMapping[index]);
+      }
+    } else {
+      // Parse Lanna desired connections
+      const lannaConnections = [
+        "getHealthy",
+        "cowork",
+        "enjoyMeals",
+        "learnFrontierTopics",
+        "findCollaborators",
+        "goExploring",
+        "party",
+        "doMentalWorkouts",
+      ];
```

Comment:
Move out of for loop, no need to redefine every time? Nit: change to `lannaInterests`

Reply:
yup, fixed